drivers/mtd_flashpage/mtd_flashpage.c

```diff
@@ -81,10 +122,47 @@ static int _write(mtd_dev_t *dev, const void *buf, uint32_t addr, uint32_t size)
     return 0;
 }
 
+static int _write_page(mtd_dev_t *dev, const void *buf, uint32_t page, uint32_t offset,
+                       uint32_t size)
+{
+    mtd_flashpage_t *super = container_of(dev, mtd_flashpage_t, base);
+
+    page += super->offset;
+    uintptr_t addr = (uintptr_t)flashpage_addr(page) + offset;
+
+    DEBUG("flashpage: write %"PRIu32" bytes from %p to %p\n", size, buf, (void *)addr);
+
+    if ((addr % FLASHPAGE_WRITE_BLOCK_ALIGNMENT) ||
+        ((uintptr_t)buf % FLASHPAGE_WRITE_BLOCK_ALIGNMENT)) {
+        uint8_t tmp[FLASHPAGE_WRITE_BLOCK_SIZE];
+
+        offset = addr % FLASHPAGE_WRITE_BLOCK_ALIGNMENT;
+        size = MIN(size, FLASHPAGE_WRITE_BLOCK_ALIGNMENT - offset);
+
+        DEBUG("flashpage: write %"PRIu32" unaligned bytes\n", size);
+
+        memcpy(&tmp[0], (uint8_t *)addr - offset, sizeof(tmp));
+        memcpy(&tmp[offset], buf, size);
+
+        flashpage_write((uint8_t *)addr - offset, tmp, sizeof(tmp));
+
+        return size;
+    }
+
+    size = MIN(flashpage_size(page) - offset, size);
```

Comment:
Let´s say `addr` and `buf` are aligned correctly and `size` is 1.
1 is not necessarily a multiple of `FLASHPAGE_WRITE_BLOCK_SIZE` as required by `flashpage_write()`.

To have all requirements checked once, maybe (`_read()` could be called from `_read_page()` and) `_write()` could be called from `_write_page()` and their return values could be asserted 0.

Reply:
I added a size check with the alignment check.